### ISSUE TITLE
Errors handling and GraphQL spec

### ISSUE THREAD

**Issue by adelcasse** [samsarahq/thunder#148]
Hello,

I'm new to this library so please excuse me if I'm out of scope or didn't understand something.

As far as I've understood, GraphQL errors are returned as strings : 

- "name.something: error message" when they are path errors (ie. we gave a name to the query and declared them with errors.New in resolvers)
- "error message" : when we use a sanitized error or if we didn't declare a name for the object.

But the GraphQL specification is quite different : http://facebook.github.io/graphql/June2018/#sec-Errors

Errors messages should look like : 

    {
      "message": "Name for character with ID 1002 could not be fetched.",
      "locations": [ { "line": 6, "column": 7 } ],
      "path": [ "hero", "heroFriends", 1, "name" ]
    }

This leads to issues with client libraries expecting the right format.

Tell me if I missed something. If so, I'm sorry.

In case I'm right, is there any interest in fixing that (which would probably be a breaking change for existing users, and Samsara in the first place) ? If so, we could probably contribute to it as we're definitely interested in using this library in our products (everything else is just perfect ;)).

**Comment by stephen** [comment]
Hi @adelcasse, you're correct in that thunder does not currently adhere to the graphql error handling mechanism. Here's some work that @changpingc did that looks similarly: https://github.com/samsara-dev/backend/pull/12845/files

We could also try to run flate (compression) ourselves on that json blob: https://golang.org/pkg/compress/flate/

**Comment by adelcasse** [comment]
Hi @stephen,

Thanks for the quick answer. I don't have access to samsara-dev/backend (probably private organization/repo ?)

**Comment by adelcasse** [comment]
@stephen I've tried something to avoid breaking changes but still be able to return valid GraphQL errors. What do you thing about that and is there interest in pursuing this implementation ?

**Comment by ccbrown** [comment]
This is kind of a big deal imo. I would call spec violations like this a deal-breaker for any API that hopes to see widespread use.

That said, I'm not a big fan of that PR as-is. Setting the location and path shouldn't be the library consumer's responsibility; it should be the library's. Imo there should simply be a flag on the schema that enables spec-compliant errors. And extensions should be facilitated via interface rather than concrete error type.

**Comment by adelcasse** [comment]
I agree. Without any answer here, we finally switched to another library.

**Comment by sijad** [comment]
@stephen can you please share your inputs about this one?
